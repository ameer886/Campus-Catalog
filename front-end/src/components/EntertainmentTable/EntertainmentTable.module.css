.EntertainmentTable {
  align-items: center;
  background-color: rgb(230, 228, 228);
  margin-top: 100px;
  font-family: 'Times New Roman', Times, serif;
}

.FilterButton {
  float: right;
  margin: 8px;
}
