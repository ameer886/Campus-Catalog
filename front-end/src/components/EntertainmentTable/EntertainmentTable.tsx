import React from 'react';
import { useState } from 'react';

import type { ColumnDefinitionType } from '../../components/GenericTable/GenericTable';
import type { EntertainmentRowType } from '../../views/Entertainments/EntertainmentsPage';
import type { IntentionallyAny } from '../../utilities';
import type { FilterPopoverOption } from '../../components/FilterPopover/FilterPopover';

import GenericTable from '../GenericTable/GenericTable';
import PaginationRelay from '../Pagination/PaginationRelay';
import PaginatedTable from '../../components/Pagination/PaginatedTable';
import FilterPopover from '../../components/FilterPopover/FilterPopover';

import styles from './EntertainmentTable.module.css';

const entertainmentTableHeaders: ColumnDefinitionType<
  EntertainmentRowType,
  keyof EntertainmentRowType
>[] = [
  {
    key: 'amen_name',
    header: 'Amenity Name',
  },
  {
    key: 'city',
    header: 'City',
  },
  {
    key: 'state',
    header: 'State',
  },
  {
    key: 'num_review',
    header: 'Reviews',
  },
  {
    key: 'pricing',
    header: 'Price',
  },
  {
    key: 'rating',
    header: 'Rating',
  },
];

// This is exported for unit tests. You should not need to import it.
export const popoverOptions: FilterPopoverOption[] = [
  {
    header: 'City',
    key: 'city',
    inputValues: [
      {
        displayStr: 'Enter City here',
        // Capitalize only the first letter
        cleanFunc: (e) =>
          e
            .trim()
            .toLowerCase()
            .replace(/^\w/, (c) => c.toUpperCase()),
      },
    ],
  },
  {
    header: 'State',
    key: 'state',
    inputValues: [
      {
        displayStr: 'Enter state here (ex: TX)',
        cleanFunc: (e) => e.trim().toUpperCase(),
      },
    ],
  },
  {
    header: 'Rating',
    key: 'rate',
    variant: 'radio',
    boxValues: [
      { value: '0', displayStr: 'Any', __checked: true },
      { value: '1.0', displayStr: '>= 1.0' },
      { value: '2.0', displayStr: '>= 2.0' },
      { value: '3.0', displayStr: '>= 3.0' },
      { value: '4.0', displayStr: '>= 4.0' },
    ],
  },
  {
    header: 'Reviews',
    key: 'reviews',
    inputValues: [
      {
        displayStr: 'Minimum # of reviews',
        min: 0,
        type: 'number',
      },
    ],
  },
  {
    header: 'Price',
    key: 'price',
    variant: 'checkbox',
    boxValues: [
      { value: '$', displayStr: '$', __checked: true },
      { value: '$$', displayStr: '$$', __checked: true },
      { value: '$$$', displayStr: '$$$', __checked: true },
      { value: '$$$$', displayStr: '$$$$', __checked: true },
      { value: '$$$$$', displayStr: '$$$$$', __checked: true },
    ],
  },
  {
    header: 'Delivery',
    key: 'delivery',
    variant: 'radio',
    boxValues: [
      { value: 'True', displayStr: 'Has Delivery' },
      { value: 'False', displayStr: 'No Delivery' },
      { value: '', displayStr: 'Any', __checked: true },
    ],
  },
  {
    header: 'Takeout',
    key: 'takeout',
    variant: 'radio',
    boxValues: [
      { value: 'True', displayStr: 'Has Takeout' },
      { value: 'False', displayStr: 'No Takeout' },
      { value: '', displayStr: 'Any', __checked: true },
    ],
  },
];

// This is an optional property to ONLY BE USED IN JEST TESTS
// PLEASE do not use testRows in production
type EntertainmentTableTestProps = {
  testRows?: Array<EntertainmentRowType>;
};

const EntertainmentTable: React.FunctionComponent<EntertainmentTableTestProps> =
  ({ testRows }: EntertainmentTableTestProps) => {
    if (testRows) {
      // Hack around queries to get a "table" that tests the basic fns
      // We need to avoid any kind of query while keeping tests functional
      // If testRows is not undefined, then this will most certainly
      // crash the front-end
      const ps = 10;
      const [filter, setFilter] = useState('');
      const [sortStr, setSortStr] = useState('NONE');
      const [page, setPage] = useState(1);

      return (
        <div className={styles.EntertainmentTable}>
          <div>
            <div className={styles.FilterButton}>
              <FilterPopover
                options={popoverOptions}
                setFilter={(e: string) => {
                  if (filter === e) return;
                  setFilter(e);
                }}
              />
            </div>
          </div>

          <GenericTable
            columnDefinitions={entertainmentTableHeaders}
            data={testRows.slice((page - 1) * ps, page * ps)}
            parentSort={setSortStr}
            parentStr={sortStr}
          />

          <PaginationRelay
            curPage={page}
            setPage={setPage}
            pageSize={ps}
            totalElements={testRows.length}
          />
        </div>
      );
    }

    const processResponse = (data: IntentionallyAny) => {
      const responseRows = data[1].amenities.map(
        (amen: IntentionallyAny) => {
          return {
            id: amen.amen_id,
            ...amen,
          };
        },
      );
      return responseRows;
    };

    return (
      <div className={styles.EntertainmentTable}>
        <PaginatedTable
          options={popoverOptions}
          processResponse={processResponse}
          model="amenities"
          columnDefinitions={entertainmentTableHeaders}
        />
      </div>
    );
  };

export default EntertainmentTable;
