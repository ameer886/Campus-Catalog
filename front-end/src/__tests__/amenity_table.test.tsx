import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import { configure, shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import { MemoryRouter } from 'react-router-dom';

import type { EntertainmentRowType } from '../views/Entertainments/EntertainmentsPage';
import EntertainmentTable from '../components/EntertainmentTable/EntertainmentTable';

configure({ adapter: new Adapter() });

const EXAMPLE_ROWS: Array<EntertainmentRowType> = [
  {
    amen_id: 621884499,
    amen_name: 'Pinballz Arcade',
    city: 'Austin',
    deliver: false,
    num_review: 516,
    pricing: 'N/A',
    rating: 4.5,
    state: 'TX',
    takeout: false,

    id: 621884499,
  },
  {
    amen_id: 485770102,
    amen_name: 'Blue Starlite Mini Urban Drive-In',
    city: 'Austin',
    deliver: false,
    num_review: 149,
    pricing: 'N/A',
    rating: 4.5,
    state: 'TX',
    takeout: false,

    id: 485770102,
  },
  {
    amen_id: 462165312,
    amen_name: 'Elephant Room',
    city: 'Austin',
    deliver: false,
    num_review: 431,
    pricing: '$$',
    rating: 4.0,
    state: 'TX',
    takeout: false,

    id: 462165312,
  },
];

jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useLocation: () => ({
    pathname: '/amenities',
  }),
}));

describe('Amenity Table Test Suite', () => {
  it('snapshot test', () => {
    const snap = shallow(
      <EntertainmentTable testRows={EXAMPLE_ROWS} />,
    );
    expect(snap).toMatchSnapshot();
  });

  it('page displays cur page and total', () => {
    render(<EntertainmentTable testRows={EXAMPLE_ROWS} />);
    const textElement = screen.getByText('Showing items 1 - 3 of 3.');
    expect(textElement).toBeInTheDocument();
  });

  it('links have correct href', () => {
    render(
      <MemoryRouter initialEntries={['/housing']}>
        <EntertainmentTable testRows={EXAMPLE_ROWS} />
      </MemoryRouter>,
    );

    for (let i = 0; i < EXAMPLE_ROWS.length; i++) {
      const aptLink = screen.getByRole('link', {
        name: EXAMPLE_ROWS[i].amen_name,
      });
      expect(aptLink).not.toBeNull();
      expect(aptLink.getAttribute('href')).toBe(
        `/amenities/${EXAMPLE_ROWS[i].id}`,
      );
    }
  });

  it('pagination works', () => {
    const rows = [];
    for (let i = 0; i < 25; i++) {
      rows.push(EXAMPLE_ROWS[0]);
    }
    render(<EntertainmentTable testRows={rows} />);

    const firstBtn = screen.getByRole('button', {
      name: 'first',
    });
    expect(firstBtn).not.toBeNull();
    const prevBtn = screen.getByRole('button', {
      name: 'prev',
    });
    expect(prevBtn).not.toBeNull();
    const nextBtn = screen.getByRole('button', {
      name: 'next',
    });
    expect(nextBtn).not.toBeNull();
    const lastBtn = screen.getByRole('button', {
      name: 'last',
    });
    expect(lastBtn).not.toBeNull();

    let textElement = screen.getByText('Showing items 1 - 10 of 25.');
    expect(textElement).toBeInTheDocument();
    expect(firstBtn).toBeDisabled();
    expect(prevBtn).toBeDisabled();

    let tbody = screen.getByRole('table').children[1];
    expect(tbody.childElementCount).toEqual(10);

    expect(nextBtn).not.toBeDisabled();
    fireEvent.click(nextBtn);
    textElement = screen.getByText('Showing items 11 - 20 of 25.');
    expect(textElement).toBeInTheDocument();

    expect(lastBtn).not.toBeDisabled();
    fireEvent.click(lastBtn);
    textElement = screen.getByText('Showing items 21 - 25 of 25.');
    expect(textElement).toBeInTheDocument();
    expect(nextBtn).toBeDisabled();
    expect(lastBtn).toBeDisabled();

    tbody = screen.getByRole('table').children[1];
    expect(tbody.childElementCount).toEqual(5);

    expect(prevBtn).not.toBeDisabled();
    fireEvent.click(prevBtn);
    textElement = screen.getByText('Showing items 11 - 20 of 25.');
    expect(textElement).toBeInTheDocument();

    expect(firstBtn).not.toBeDisabled();
    fireEvent.click(firstBtn);
    textElement = screen.getByText('Showing items 1 - 10 of 25.');
    expect(textElement).toBeInTheDocument();
  });
});
