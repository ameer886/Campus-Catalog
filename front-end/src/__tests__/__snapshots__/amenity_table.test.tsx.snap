// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Amenity Table Test Suite snapshot test 1`] = `ShallowWrapper {}`;
